pkg/controllers/cspc-controller/sync.go

```diff
@@ -71,6 +77,32 @@ func (c *Controller) sync(cspc *cstor.CStorPoolCluster, cspiList *cstor.CStorPoo
 		return nil
 	}
 
+	cspcGot, err = c.reconcileVersion(cspcGot)
+	if err != nil {
+		message := fmt.Sprintf("Failed to upgrade cspc to %s version: %s",
+			cspcGot.VersionDetails.Desired,
+			err.Error())
+		klog.Errorf("failed to upgrade cspc %s:%s", cspcGot.Name, err.Error())
+		c.recorder.Event(cspc, corev1.EventTypeWarning, "FailedUpgrade", message)
+		cspcGot.VersionDetails.Status.SetErrorStatus(
+			"Failed to reconcile cspc version",
+			err,
+		)
+		_, err = c.clientset.CstorV1().CStorPoolClusters(cspcGot.Namespace).Update(cspcGot)
+		if err != nil {
+			klog.Errorf("failed to update versionDetails status for cspc %s:%s", cspcGot.Name, err.Error())
+		}
+		return nil
+	}
+
+	if ok, reason := c.ShouldReconcile(*cspc); !ok {
```

Comment:
It should move below the code
```go
	if !cspcGot.DeletionTimestamp.IsZero() {
}
```

Reply:
When the controller is upgraded then reconcile version will never be triggered after patching the VersionDetails.DesiredVersion to the upgraded version as the ShouldReconcile function checks for CurrentVersion. Once the reconcileVersion is successful the CurrentVersion on the CR will be changed to the new version.

Reply:
I got it but we can delete the resource irrespective of current version so need to move below https://github.com/openebs/cstor-operators/blob/bc2114c8698a27a602fb55f419618de7e9fe16d4/pkg/controllers/cspc-controller/sync.go#L82

(Or)

[This snippet](https://github.com/openebs/cstor-operators/blob/bc2114c8698a27a602fb55f419618de7e9fe16d4/pkg/controllers/cspc-controller/sync.go#L82) should move before any function.

Reply:
Got it! Nice Catch :+1: